File: analyze.py
import re
import os
from collections import Counter

folder = r"C:\Users\A1\Desktop\Millione\المليون"
files = ['1.txt', '2.txt', '3.txt', '4.txt', '5.txt']

all_ids = []
all_questions = []

print("=== تحليل ملفات الأسئلة ===\n")

for fname in files:
    fpath = os.path.join(folder, fname)
    content = open(fpath, encoding='utf-8', errors='ignore').read()
    
    ids = re.findall(r'"id"\s*:\s*(\d+)', content)
    ids = [int(x) for x in ids]
    
    questions = re.findall(r'"question"\s*:\s*"([^"]+)"', content)
    
    print(f"{fname}: {len(ids)} سؤال | IDs من {min(ids) if ids else 'N/A'} إلى {max(ids) if ids else 'N/A'}")
    
    all_ids.extend(ids)
    all_questions.extend(questions)

print(f"\n{'='*40}")
print(f"الإجمالي الكلي (كل الملفات): {len(all_ids)}")

# Duplicate IDs
id_counter = Counter(all_ids)
dup_ids = {k: v for k, v in id_counter.items() if v > 1}
print(f"IDs فريدة: {len(id_counter)}")
print(f"IDs مكررة: {len(dup_ids)} معرف")

# Duplicate Questions (text)
q_counter = Counter(all_questions)
dup_questions = {k: v for k, v in q_counter.items() if v > 1}
print(f"\nنصوص أسئلة فريدة: {len(q_counter)}")
print(f"نصوص أسئلة مكررة: {len(dup_questions)} سؤال")

if dup_questions:
    print(f"\nأمثلة على الأسئلة المكررة (أول 10):")
    for q, count in list(dup_questions.items())[:10]:
        print(f"  [x{count}] {q[:70]}")

print(f"\n{'='*40}")
print(f"الأسئلة الفريدة الحقيقية (بدون تكرار): {len(q_counter)}")
print(f"نسبة التكرار: {(len(all_questions) - len(q_counter)) / len(all_questions) * 100:.1f}%")

File: analyze.js
const fs = require('fs');
const path = require('path');

const folder = path.join(__dirname, 'المليون');
const files = ['1.txt', '2.txt', '3.txt', '4.txt', '5.txt'];

let allIds = [];
let allQuestions = [];

console.log('=== تحليل ملفات الأسئلة ===\n');

files.forEach(file => {
    const filePath = path.join(folder, file);
    const content = fs.readFileSync(filePath, 'utf8');
    
    // Extract all "id": NUMBER patterns
    const idMatches = [...content.matchAll(/"id"\s*:\s*(\d+)/g)];
    const ids = idMatches.map(m => parseInt(m[1]));
    
    // Extract question texts to check text-level duplicates
    const qMatches = [...content.matchAll(/"question"\s*:\s*"([^"]+)"/g)];
    const questions = qMatches.map(m => m[1].trim());
    
    console.log(`${file}: ${ids.length} سؤال (IDs: ${ids[0]} -> ${ids[ids.length-1]})`);
    
    allIds.push(...ids);
    allQuestions.push(...questions);
});

console.log('\n=== النتائج الإجمالية ===');
console.log(`إجمالي الأسئلة (كل الملفات): ${allIds.length}`);

// Check duplicate IDs
const idCount = {};
allIds.forEach(id => { idCount[id] = (idCount[id] || 0) + 1; });
const dupIds = Object.entries(idCount).filter(([id, count]) => count > 1);
const uniqueIds = Object.keys(idCount).length;

console.log(`IDs فريدة: ${uniqueIds}`);
console.log(`IDs مكررة: ${dupIds.length} معرف`);

if (dupIds.length > 0 && dupIds.length <= 30) {
    console.log('\nالـ IDs المكررة:');
    dupIds.forEach(([id, count]) => console.log(`  ID ${id}: مكرر ${count} مرة`));
} else if (dupIds.length > 30) {
    console.log(`\nأول 20 ID مكرر:`);
    dupIds.slice(0, 20).forEach(([id, count]) => console.log(`  ID ${id}: مكرر ${count} مرة`));
}

// Check duplicate QUESTIONS (text-level)
const qCount = {};
allQuestions.forEach(q => { qCount[q] = (qCount[q] || 0) + 1; });
const dupQuestions = Object.entries(qCount).filter(([q, count]) => count > 1);

console.log(`\nنصوص أسئلة فريدة: ${Object.keys(qCount).length}`);
console.log(`نصوص أسئلة مكررة: ${dupQuestions.length} سؤال`);

if (dupQuestions.length > 0 && dupQuestions.length <= 20) {
    console.log('\nأسئلة مكررة (نص):');
    dupQuestions.forEach(([q, count]) => console.log(`  [x${count}] ${q.substring(0, 60)}...`));
} else if (dupQuestions.length > 20) {
    console.log(`\nأول 10 أسئلة مكررة (نص):`);
    dupQuestions.slice(0, 10).forEach(([q, count]) => console.log(`  [x${count}] ${q.substring(0, 60)}`));
}

console.log('\n=== ملخص ===');
console.log(`الإجمالي الحقيقي بعد إزالة التكرار: ${Object.keys(qCount).length} سؤال فريد`);
